### PR TITLE
enhancement(prometheus_scrape source): update bucket size to u64

### LINKED ISSUE #12043: Prometheus Scrape fails when encountering a _bucket value greater than MAX:u32
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

Prometheus Scrape fails when encountering a _bucket value greater than MAX:u32

When scraping a concourse metrics endpoint using vector, concourse presents the following histogram metric:

```
go_gc_heap_allocs_by_size_bytes_total_bucket{le="24.999999999999996"} 1.8939392877e+10
```

Vector, in turn presents the following error message about this line: 
```
Mar 31 14:21:08 ip-10-1-1-200 vector[76587]: 2022-03-31T14:21:08.003924Z ERROR source{component_kind="source" component_id=collect_concourse_metrics component_type=prometheus_scrape component_name=collect_concourse_metrics}: vector::internal_events::prometheus: Parsing error. url=http://localhost:9391/metrics error=ValueOutOfRange { value: 18939392877.0 } error_type="parse_failed" stage="processing" internal_log_rate_secs=1
```

Which I believe I have traced down to these bits of code:

https://github.com/vectordotdev/vector/blob/a398504fe21ff64a065f97b203539ce986e4e889/lib/prometheus-parser/src/lib.rs#L169-L178

https://github.com/vectordotdev/vector/blob/a398504fe21ff64a065f97b203539ce986e4e889/lib/prometheus-parser/src/lib.rs#L238-L244

And 18939392877 is, indeed, larger than 4294967295. But I feel like is a valid value for a metric, is it not? There shouldn't really be a limit?

I don't actually care about this particular metric, I even discard it in my transforms, but as a result of this parse failure, vector seems to discard the entire scrape and nothing is passed through to my transforms or sinks.

### Configuration

```text
sources:
  collect_concourse_metrics:
    type: prometheus_scrape
    endpoints:
    - http://localhost:9391/metrics
    scrape_interval_secs: 10
```

### Version

vector 0.20.0 (x86_64-unknown-linux-gnu 2a706a3 2022-02-11)

### Debug Output

_No response_

### Example Data

_No response_

### Additional Context

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #12043 

This updates the Histogram bucket size to store `u64` rather than `u32`.

A number of issues have come up from this.

1. `DDSketch` stores u32, this means in `insert_interpolate_buckets` we now check that the count fits into `u32` and error if it doesn't. This means `DatadogMetricsNormalizer` will drop the metric if it is > `u32::MAX`. Is this an ok thing to do?
~~2. From what I gather [InfluxDB can only handle `u32`](https://github.com/influxdata/influxdb/issues/7801#issuecomment-466801839), unless you create a custom build. What would be the best way to handle this? Send u64 and it's up to the user to ensure it works ok, send u32 and drop > `u32::MAX` metrics or make it a configurable option?~~

I haven't done a complete survey yet through all our sinks to check if all the services can handle u64.

### CODE REVIEW

**Review by jszwedko — COMMENTED** [review]
Nice! I was expecting this to take much longer. I still need to review in detail, but just a note that we should make sure to put together a note for the upgrade guide since users could run into issues with this change. We should note:

* That disk buffers should be backed up if you want to be able to rollback since new disk buffer entries may not be readable by older Vector versions
* Users sending events between Vector instances should upgrade the consumers first, then the producers

**Review by bruceg — APPROVED** [review]

**Review by tobz — CHANGES_REQUESTED** [review]

**Comment by tobz** [comment]
> > * `DDSketch` stores u32, this means in `insert_interpolate_buckets` we now check that the count fits into `u32` and error if it doesn't. This means `DatadogMetricsNormalizer` will drop the metric if it is > `u32::MAX`. Is this an ok thing to do?
> 
> I think Toby had ideas here around splitting the bucket?

At a high-level, we force all aggregated histograms into incremental versions, so the only way to violate bucket counts being less than or equal to `u32::MAX` would imply a bucket count growing by `u32::MAX` or more from one update to the next... not impossible, but practically won't ever happen.

As far as splitting the bins in the sketch: I double checked, and it's already implemented internally and will happen automatically... but the problem is that bin counts are `u16`. This means we would need 2^16 bins to represent a single histogram bucket having even a count of `u32::MAX` without this PR, or 2^48 bins to represent a count of `u64::MAX`. This is already not going to be possible with the sketch defaults we use, which limit us to a maximum of 4096 bins per sketch. Practically, this still limits us to a single bucket in an aggregated histogram changing its count by ~268 million or less to avoid an overflow..

It'd be nice if we had a more compile-time/type-level way to do all of this, but I think the best we can do is maybe log an error when we encounter such a weird incremental aggregated histogram in the DD metrics normalizer. 🤷🏻

**Comment by StephenWakely** [comment]
> Do you know how InfluxDB responds when you send a u64?

Ok have tested this. It seems that (at least the Docker builds) are capable of receiving u64 fine. If I do send a count that is definitely out of range, Influx will return a 400 error.

**Comment by jszwedko** [comment]
> > Do you know how InfluxDB responds when you send a u64?
> 
> Ok have tested this. It seems that (at least the Docker builds) are capable of receiving u64 fine. If I do send a count that is definitely out of range, Influx will return a 400 error.

Cool, it seems sensible to me to just pass through then.

File: lib/prometheus-parser/src/line.rs [hunk 1]
```diff
@@ -560,6 +560,28 @@ mod test {
...
+        // let input = wrap(&(u64::MAX - 1).to_string());
+        // let (left, r) = Metric::parse_value(&input).unwrap();
+        // assert_eq!(left, tail);
+        // assert_eq!(r as u64, u64::MAX - 1);
+
+        let input = wrap(&(u32::MAX as u64 + 1).to_string());
+        let (left, r) = Metric::parse_value(&input).unwrap();
+        assert_eq!(left, tail);
+        assert_eq!(r as u64, u32::MAX as u64 + 1);
+
```
Comment (tobz) [inline lib/prometheus-parser/src/line.rs:584]: I think you can reduce this section down to a smaller comment that explains that converting from an integer to a floating-point number will start to lose precision once the number is bigger than 2^N, where N is the number of bits in the mantissa (53 bits for an `f64`).

In addition, or alternatively, you could add an assertion that we handle 2^N-1 and 2^N without loss, but that 2^N+1 rounds down to 2^N, to show we're providing a maximally correct conversion where possible.

**Review by tobz — APPROVED** [review]
Yeah, this looks reasonable enough to me. 👍🏻 

Left one small non-blocking nit.

**Pull request merged**